### PR TITLE
Unit test for trust manager supporting local JRE key store

### PR DESCRIPTION
**Pull request by michaelqi793**

### CODE REVIEW

File: sdk/keyvault/azure-security-keyvault-jca/src/test/java/com/azure/security/keyvault/jca/JREKeystoreTest.java [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class JREKeystoreTest {
+    private static final String fileSep = File.separator;
+    private static final String defaultStorePath =
+        PrivilegedActionImpl.privilegedGetProperty("java.home", "") +
+            fileSep + "lib" + fileSep + "security";
+    private static final String defaultStore =
+        defaultStorePath + fileSep + "cacerts";
```
Comment (chenrujun) [inline sdk/keyvault/azure-security-keyvault-jca/src/test/java/com/azure/security/keyvault/jca/JREKeystoreTest.java:39]: Do not break line if it not exceed 120 width.

File: sdk/keyvault/azure-security-keyvault-jca/src/test/java/com/azure/security/keyvault/jca/JREKeystoreTest.java [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+    private static final String jsseDefaultStore =
+        defaultStorePath + fileSep + "jssecacerts";
```
Comment (chenrujun) [inline sdk/keyvault/azure-security-keyvault-jca/src/test/java/com/azure/security/keyvault/jca/JREKeystoreTest.java:41]: Same here.

File: sdk/keyvault/azure-security-keyvault-jca/src/test/java/com/azure/security/keyvault/jca/JREKeystoreTest.java [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+    private PrivateKeyStrategyImpl aliasStrategy = new PrivateKeyStrategyImpl();
+    private static final String keyStorePassword = PrivilegedActionImpl.privilegedGetProperty("javax.net.ssl.keyStorePassword", "changeit");
+    // private static final String keyStoreType = PrivilegedActionImpl.privilegedGetProperty("javax.net.ssl.keyStoreType", KeyStore.getDefaultType());
+    //private static final String keyStoreProvider = PrivilegedActionImpl.privilegedGetProperty("javax.net.ssl.keyStoreProvider", "");
```
Comment (chenrujun) [inline sdk/keyvault/azure-security-keyvault-jca/src/test/java/com/azure/security/keyvault/jca/JREKeystoreTest.java:45]: This is used to debug?

File: sdk/keyvault/azure-security-keyvault-jca/src/test/java/com/azure/security/keyvault/jca/JREKeystoreTest.java [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+import java.security.spec.PKCS8EncodedKeySpec;
+import java.util.Base64;
+import java.util.Map;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class JREKeystoreTest {
```
Comment (chenrujun) [inline sdk/keyvault/azure-security-keyvault-jca/src/test/java/com/azure/security/keyvault/jca/JREKeystoreTest.java:33]: How about rename to `KeyVaultTrustManagerFactoryProviderTest`?

File: sdk/keyvault/azure-security-keyvault-jca/src/test/java/com/azure/security/keyvault/jca/JREKeystoreTest.java [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+                    logger.log(Level.SEVERE, "Server can't write to the client socket.", ioe);
+                }
+            }
+        });
+        serverThread.start();
+
+        /*
+         * Setup client side
+         */
+
```
Comment (chenrujun) [inline sdk/keyvault/azure-security-keyvault-jca/src/test/java/com/azure/security/keyvault/jca/JREKeystoreTest.java:140]: Please re-arrange the empty lines:
1. Delete unnecessary empty line.
2. Do not put 2 empty line together in normal case.

File: sdk/keyvault/azure-security-keyvault-jca/src/test/java/com/azure/security/keyvault/jca/JREKeystoreTest.java [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+        Security.insertProviderAt(new KeyVaultTrustManagerFactoryProvider(), 1);
+        KeyStore keyStore = KeyStore.getInstance(KeyStore.getDefaultType());
+        try (final FileInputStream inStream = new FileInputStream(getKeyStoreFile())) {
+            keyStore.load(inStream, keyStorePassword.toCharArray());
+        }
+        keyStore.setEntry(aliasStrategy.chooseAlias(null, null), new KeyStore.PrivateKeyEntry(getKey(), getCertChain()), new KeyStore.PasswordProtection(keyPassword.toCharArray()));
+        /*
+         * Setup server side.
+         */
```
Comment (chenrujun) [inline sdk/keyvault/azure-security-keyvault-jca/src/test/java/com/azure/security/keyvault/jca/JREKeystoreTest.java:113]: Could you please wrap these into a small method? Or into a new class.

File: sdk/keyvault/azure-security-keyvault-jca/src/test/java/com/azure/security/keyvault/jca/JREKeystoreTest.java [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+    private static class PrivateKeyStrategyImpl implements PrivateKeyStrategy {
+        @Override
+        public String chooseAlias(Map<String, PrivateKeyDetails> map, Socket socket) {
+            return "a-humble-key-entry-alias";
+        }
+
+    }
+
+
+    private static class PrivilegedActionImpl implements PrivilegedAction<String> {
```
Comment (chenrujun) [inline sdk/keyvault/azure-security-keyvault-jca/src/test/java/com/azure/security/keyvault/jca/JREKeystoreTest.java:247]: Can we do this with less code?

File: sdk/keyvault/azure-security-keyvault-jca/src/test/java/com/azure/security/keyvault/jca/JREKeystoreTest.java [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+        }
+
+        private PrivilegedActionImpl(String theProp, String defaultVal) {
+            this.theProp = theProp;
+            this.defaultVal = defaultVal;
+
+        }
+
+        public String run() {
+            String value = System.getProperty(theProp);
```
Comment (chenrujun) [inline sdk/keyvault/azure-security-keyvault-jca/src/test/java/com/azure/security/keyvault/jca/JREKeystoreTest.java:274]: `System.getProperty` need to be `doPrivileged`?

**Review by chenrujun — CHANGES_REQUESTED** [review]

**Comment by chenrujun** [comment]
Hi, @michaelqi793 , please fix the pipeline failure.

File: sdk/keyvault/azure-security-keyvault-jca/src/test/java/com/azure/security/keyvault/jca/JREKeystoreTest.java (hunk 8 above)
Comment (michaelqi793) [reply sdk/keyvault/azure-security-keyvault-jca/src/test/java/com/azure/security/keyvault/jca/JREKeystoreTest.java:274]: Looks like yes if a security manager is installed.

File: sdk/keyvault/azure-security-keyvault-jca/src/test/java/com/azure/security/keyvault/jca/JREKeystoreTest.java [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+import java.util.Map;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class JREKeystoreTest {
+    private static final String fileSep = File.separator;
+    private static final String defaultStorePath = PrivilegedActionImpl.privilegedGetProperty("java.home", "") + fileSep + "lib" + fileSep + "security";
```
Comment (stliu) [inline sdk/keyvault/azure-security-keyvault-jca/src/test/java/com/azure/security/keyvault/jca/JREKeystoreTest.java:35]: refactor this to use java8's File/Path api?

File: sdk/keyvault/azure-security-keyvault-jca/src/test/java/com/azure/security/keyvault/jca/JREKeystoreTest.java [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+        try (final FileInputStream inStream = new FileInputStream(getKeyStoreFile())) {
+            keyStore.load(inStream, keyStorePassword.toCharArray());
+        }
+        keyStore.setEntry(aliasStrategy.chooseAlias(null, null), new KeyStore.PrivateKeyEntry(getKey(), getCertChain()), new KeyStore.PasswordProtection(keyPassword.toCharArray()));
+
+        /*
+         * Setup server side.
+         */
+        int randomPort = ThreadLocalRandom.current().nextInt(49452, 52570);
+        logger.log(Level.INFO, "The port used is " + randomPort);
```
Comment (stliu) [inline sdk/keyvault/azure-security-keyvault-jca/src/test/java/com/azure/security/keyvault/jca/JREKeystoreTest.java:103]: debug level

File: sdk/keyvault/azure-security-keyvault-jca/src/test/java/com/azure/security/keyvault/jca/JREKeystoreTest.java (hunk 9 above)
Comment (michaelqi793) [reply sdk/keyvault/azure-security-keyvault-jca/src/test/java/com/azure/security/keyvault/jca/JREKeystoreTest.java]: Done

File: sdk/keyvault/azure-security-keyvault-jca/src/test/java/com/azure/security/keyvault/jca/JREKeystoreTest.java (hunk 4 above)
Comment (michaelqi793) [reply sdk/keyvault/azure-security-keyvault-jca/src/test/java/com/azure/security/keyvault/jca/JREKeystoreTest.java]: Changed~

**Review by backwind1233 — CHANGES_REQUESTED** [review]
please fix the pipeline build error

**Comment by chenrujun** [comment]
Hi, @michaelqi793 , this PR can be closed, right?

**Comment by michaelqi793** [comment]
> Hi, @michaelqi793 , this PR can be closed, right?

Yes!

**Pull request closed without merging**